salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+            self._rmq_connection_wrapper.consume_reply(mark_future, reply_queue_name=message_correlation_id)
+
+            try:
+                ret = yield future
+            except Exception as err:  # pylint: disable=broad-except
+                del self.send_queue[0]
+                continue
+            del self.send_queue[0]
+            self.send_future_map.pop(message, None)
+            self.remove_message_timeout(message)
+
+    def remove_message_timeout(self, message):
+        if message not in self.send_timeout_map:
+            return
+        timeout = self.send_timeout_map.pop(message, None)
+        if timeout is not None:
+            # Hasn't been already timedout
+            self.io_loop.remove_timeout(timeout)
+
+    def timeout_message(self, message):
+        """
+        Handle a message timeout by removing it from the sending queue
+        and informing the caller
+
+        :raises: SaltReqTimeoutError
+        """
+        future = self.send_future_map.pop(message, None)
+        # In a race condition the message might have been sent by the time
+        # we're timing it out. Make sure the future is not None
+        if future is not None:
+            del self.send_timeout_map[message]
+            if future.attempts < future.tries:
+                future.attempts += 1
+                log.debug(
+                    "SaltReqTimeoutError, retrying. (%s/%s)",
+                    future.attempts,
+                    future.tries,
+                )
+                self.send(
+                    message, timeout=future.timeout, tries=future.tries, future=future,
+                )
+
+            else:
+                future.set_exception(SaltReqTimeoutError("Message timed out"))
+
+    def send(self, message, timeout=None, tries=3, future=None, callback=None, raw=False):
+        """
+        Return a future which will be completed when the message has a response
+        """
+        if future is None:
+            future = salt.ext.tornado.concurrent.Future()
+            future.tries = tries
+            future.attempts = 0
+            future.timeout = timeout
+            # if a future wasn't passed in, we need to serialize the message
+            message = self.serial.dumps(message)
+        if callback is not None:
+
+            def handle_future(future):
+                response = future.result()
+                self.io_loop.add_callback(callback, response)
+
+            future.add_done_callback(handle_future)
+        # Add this future to the mapping
+        self.send_future_map[message] = future
+
+        if self.opts.get("detect_mode") is True:
+            timeout = 1
+
+        if timeout is not None:
+            send_timeout = self.io_loop.call_later(
+                timeout, self.timeout_message, message
+            )
+            self.send_timeout_map[message] = send_timeout
+
+        if len(self.send_queue) == 0:
```

Comment:
**suggestion (non-blocking)** same Truthy suggestion here - things with stuff in them are true, so `if not self.send_queue` would work here.